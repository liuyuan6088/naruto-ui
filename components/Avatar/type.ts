
import { ICommonProps, tuple } from '../utils/type';

export const AvatarSizes = tuple('large', 'default', 'small');
export type AvatarSize = (typeof AvatarSizes)[number];

export const AvatarShape = tuple('circle', 'square');
export type AvatarShape = (typeof AvatarShape)[number];

export interface IAvatarProps extends ICommonProps {
  children?: React.ReactNode;
  // 设置头像的图标类型，参考 Icon 组件
  icon?: string;
  // 设置头像的大小
  size?: number | AvatarSize;
  // 指定头像的形状
  shape?: AvatarShape;
  // 图片类头像的资源地址
  src?: string;
  // 设置图片类头像响应式资源地址
  srcSet?: string;
  // 图像无法显示时的替代文本
  alt?: string;
  // 图片加载失败的事件，返回 false 会关闭组件默认的 fallback 行为
  onError?: () => boolean;
  // 点击事件
  onClick?: (e: React.MouseEvent) => void;
}