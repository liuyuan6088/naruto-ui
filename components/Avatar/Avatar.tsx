import * as React from 'react'
import { IAvatarProps } from './type'
import cx from 'classnames'
import { primaryName } from '../utils/constant'
import Icon from '../Icon'
import './style/index.less'

const { useRef, useState, useEffect } = React

const defaultProps: IAvatarProps = {
  prefixCls: `${primaryName}-avatar`,
  shape: 'circle',
  srcSet: '',
  alt: '',
  onError: () => false
}

const renderChildren = (
  { icon, children, size, src, srcSet, alt, onError, prefixCls }: IAvatarProps,
  childrenRef: React.MutableRefObject<HTMLSpanElement>,
  scale: number
) => {
  let childrenNode = children
  if (icon) {
    const sizeStyle =
      typeof size === 'number'
        ? {
            fontSize: size / 2
          }
        : {}
    childrenNode = <Icon style={{ ...sizeStyle }} type={icon} />
  } else if (src) {
    childrenNode = <img src={src} srcSet={srcSet} alt={alt} onError={onError} />
  } else if (typeof children === 'string') {
    const transformString = `scale(${scale}) translateX(-50%)`
    const childrenStyle: React.CSSProperties =
      scale === 1
        ? {}
        : {
            msTransform: transformString,
            WebkitTransform: transformString,
            transform: transformString
          }
    const classes = scale === 1 ? '' : `${prefixCls}-string`
    return (
      <span className={classes} style={{ ...childrenStyle }} ref={childrenRef}>
        {children}
      </span>
    )
  }
  return childrenNode
}

const getScale = (
  avatarRef: React.MutableRefObject<HTMLSpanElement>,
  childrenRef: React.MutableRefObject<HTMLSpanElement>
): number => {
  if (childrenRef && childrenRef.current) {
    const childrenNode = childrenRef.current
    const childrenWidth = childrenNode.offsetWidth
    // 外层盒子宽度
    const avatarWidth = avatarRef.current.getBoundingClientRect().width
    if (avatarWidth - 8 < childrenWidth) {
      return (avatarWidth - 8) / childrenWidth
    } else {
      return 1
    }
  }
  return 1
}

const onlyChild = (props: IAvatarProps) => !['icon', 'src'].some(e => e in props)

const Avatar: React.SFC<IAvatarProps> = props => {
  const { prefixCls, className, style, size, shape, src, onClick, children } = props

  const [scale, setScale] = useState<number>(1)

  const avatarRef = useRef<HTMLSpanElement>(null)
  const childrenRef = useRef<HTMLSpanElement>(null)

  useEffect(() => {
    if (typeof children === 'string' && onlyChild(props)) {
      const newScale = getScale(avatarRef, childrenRef)
      setScale(newScale)
    }
  }, [children])

  const classes = cx(
    prefixCls,
    {
      [`${prefixCls}-${size}`]: !!size && typeof size === 'string',
      [`${prefixCls}-${shape}`]: !!shape,
      [`${prefixCls}-image`]: !!src
    },
    className
  )

  const sizeStyle =
    typeof size === 'number'
      ? {
          width: size,
          height: size,
          lineHeight: `${size}px`,
          fontSize: size / 2
        }
      : {}

  return (
    <span ref={avatarRef} onClick={onClick} className={classes} style={{ ...style, ...sizeStyle }}>
      {renderChildren(props, childrenRef, scale)}
    </span>
  )
}

Avatar.defaultProps = defaultProps

export default Avatar
