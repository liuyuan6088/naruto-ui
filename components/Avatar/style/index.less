@import '../../styles/theme/default.less';

@avatar-prefix-cls: ~"@{primary-name}-avatar";

.getBox(@px) {
  width: @px;
  height: @px;
  line-height: @px;
}

.@{avatar-prefix-cls} {
  background: #ccc;
  display: inline-block;
  overflow: hidden;
  color: #fff;
  white-space: nowrap;
  text-align: center;
  vertical-align: middle;
  margin: 0;
  padding: 0;
  cursor: pointer;
  position: relative;
  transition: all .3s ease;
  .getBox(@avatar-default);

  &-large {
    .getBox(@avatar-lg);
    > svg {
      font-size: @avatar-lg / 2;
    }
  }

  &-small {
    .getBox(@avatar-sm);
    > svg {
      font-size: 14px;
    }
  }

  &-circle {
    border-radius: 50%;
  }

  &-square {
    border-radius: @bordor-radius;
  }

  &-image {
    img {
      width: 100%;
      height: 100%;
    }
  }

  &-string {
    position: absolute;
    left: 50%;
    transform-origin: 0 center;
  }
}