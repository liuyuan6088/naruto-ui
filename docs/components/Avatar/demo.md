---
imports:
  import Demo from './demo.tsx';
  import DemoCode from '!raw-loader!./demo.tsx';

---

## Avatar 头像

用来代表用户或事物，支持图片、图标或字符展示。

## 使用头像

:::demo
  <CodeBox code={DemoCode}>
    <Demo />
  </CodeBox>
:::

## Props
| 参数     | 说明     | 类型   | 可选值 | 默认值 |
| -------- | -------- | ------ | ------ | ------ |
| icon | 设置头像的图标类型 | string | - | - |
| size | 设置头像的大小 | number / `large` /  `default` / `small` | - | - |
| shape | 指定头像的形状 | string | `circle` / `square` | `circle` |
| src | 图片类头像的资源地址 | string | - | - |
| srcSet | 设置图片类头像响应式资源地址 | string | - | - |
| alt | 图像无法显示时的替代文本 | string | - | - |
| onError | 图片加载失败的事件，返回 false 会关闭组件默认的 fallback 行为 | () => boolean | - | - |
| onClick | 点击事件 | (e: React.MouseEvent) => void | - | - |

