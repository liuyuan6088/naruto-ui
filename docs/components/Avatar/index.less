.demo-avatar {
  > span {
    margin-right: 20px;
  }
}