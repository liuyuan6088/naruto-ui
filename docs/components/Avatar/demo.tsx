import * as React from 'react';
import { Avatar, Button } from 'components';
import './index.less';

const { useState } = React;

const userList = ['U', 'Lucy', 'Tom', 'Edward'];
const colorList = ['#f56a00', '#7265e6', '#ffbf00', '#00a2ae'];

const Demo: React.FC = () => {

  const [color, setColor] = useState<string>(colorList[0]);
  const [user, setUser] = useState<string>(userList[0]);

  const changeUser = () => {
    const index = userList.indexOf(user);
    setUser(index < userList.length - 1 ? userList[index + 1] : userList[0]);
    setColor(index < colorList.length - 1 ? colorList[index + 1] : colorList[0]);
  }

  return (
    <React.Fragment>

      <p>头像有三种尺寸, 有两种形状可选</p>
      <div className='demo-avatar'>
        <Avatar size={80} icon='user'>dada</Avatar>
        <Avatar size='large' icon='user'>dada</Avatar>
        <Avatar icon='user'>dada</Avatar>
        <Avatar size='small' icon='user'>dada</Avatar>

        <Avatar shape='square' size={80} icon='user'>dada</Avatar>
        <Avatar shape='square' size='large' icon='user'>dada</Avatar>
        <Avatar shape='square' icon='user'>dada</Avatar>
        <Avatar shape='square' size='small' icon='user'>dada</Avatar>
      </div>

      <p>支持三种类型：图片、Icon 以及字符，其中 Icon 和字符型可以自定义图标颜色及背景色</p>
      <div className='demo-avatar'>
        <Avatar src='https://zos.alipayobjects.com/rmsportal/ODTLcjxAfvqbxHnVXCYX.png'>dada</Avatar>
        <Avatar>dada</Avatar>
        <Avatar icon='user'>dada</Avatar>
        <Avatar style={{ color: '#f56a00', backgroundColor: '#fde3cf' }}>U</Avatar>
        <Avatar style={{ backgroundColor: '#87d068' }} icon="user" />
      </div>

      <p>自动调整字符大小</p>
      <div>
        <Avatar style={{ backgroundColor: color, verticalAlign: 'middle' }} size="large">
          {user}
        </Avatar>
        <Button
          size="small"
          style={{ marginLeft: 16, verticalAlign: 'middle' }}
          onClick={changeUser}
        >
          Change
        </Button>
      </div>

    </React.Fragment>
  )
}

export default Demo;
