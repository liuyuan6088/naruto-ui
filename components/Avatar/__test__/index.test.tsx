import * as React from 'react'
import { mount, render } from 'enzyme'
import Avatar from '../index'

describe('Render Avatar', () => {
  it('render default', () => {
    const wrapper = render(
      <div>
        <Avatar src="https://zos.alipayobjects.com/rmsportal/ODTLcjxAfvqbxHnVXCYX.png" />
        <Avatar icon="user" />
        <Avatar style={{ color: '#f56a00', backgroundColor: '#fde3cf' }}>U</Avatar>
        <Avatar style={{ backgroundColor: '#87d068' }} icon="user" />
      </div>
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('Render long string correctly', () => {
    const wrapper = mount(<Avatar>TestString</Avatar>)
    const children = wrapper.find('.naruto-avatar-string')
    expect(children.length).toBe(1)
  })
})
