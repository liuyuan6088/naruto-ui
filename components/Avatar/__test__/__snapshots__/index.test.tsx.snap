// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Avatar render default 1`] = `
initialize {
  "0": Object {
    "attribs": Object {},
    "children": Array [
      Object {
        "attribs": Object {
          "class": "naruto-avatar naruto-avatar-circle naruto-avatar-image",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "alt": "",
              "src": "https://zos.alipayobjects.com/rmsportal/ODTLcjxAfvqbxHnVXCYX.png",
              "srcset": "",
            },
            "children": Array [],
            "name": "img",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "alt": undefined,
              "src": undefined,
              "srcset": undefined,
            },
            "x-attribsPrefix": Object {
              "alt": undefined,
              "src": undefined,
              "srcset": undefined,
            },
          },
        ],
        "name": "span",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Object {
          "attribs": Object {
            "class": "naruto-avatar naruto-avatar-circle",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "class": "naruto-icon naruto-icon-user",
              },
              "children": Array [
                Object {
                  "attribs": Object {
                    "href": "#icon-user",
                  },
                  "children": Array [],
                  "name": "use",
                  "namespace": "http://www.w3.org/2000/svg",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "href": "http://www.w3.org/1999/xlink",
                  },
                  "x-attribsPrefix": Object {
                    "href": "xlink",
                  },
                },
              ],
              "name": "svg",
              "namespace": "http://www.w3.org/2000/svg",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
          ],
          "name": "span",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": Object {
            "attribs": Object {
              "class": "naruto-avatar naruto-avatar-circle",
              "style": "color:#f56a00;background-color:#fde3cf",
            },
            "children": Array [
              Object {
                "attribs": Object {
                  "class": "",
                },
                "children": Array [
                  Object {
                    "data": "U",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "text",
                  },
                ],
                "name": "span",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                },
              },
            ],
            "name": "span",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": Object {
              "attribs": Object {
                "class": "naruto-avatar naruto-avatar-circle",
                "style": "background-color:#87d068",
              },
              "children": Array [
                Object {
                  "attribs": Object {
                    "class": "naruto-icon naruto-icon-user",
                  },
                  "children": Array [
                    Object {
                      "attribs": Object {
                        "href": "#icon-user",
                      },
                      "children": Array [],
                      "name": "use",
                      "namespace": "http://www.w3.org/2000/svg",
                      "next": null,
                      "parent": [Circular],
                      "prev": null,
                      "type": "tag",
                      "x-attribsNamespace": Object {
                        "href": "http://www.w3.org/1999/xlink",
                      },
                      "x-attribsPrefix": Object {
                        "href": "xlink",
                      },
                    },
                  ],
                  "name": "svg",
                  "namespace": "http://www.w3.org/2000/svg",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "class": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "class": undefined,
                  },
                },
              ],
              "name": "span",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": [Circular],
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
                "style": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
                "style": undefined,
              },
            },
            "parent": [Circular],
            "prev": [Circular],
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
              "style": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
              "style": undefined,
            },
          },
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
      Object {
        "attribs": Object {
          "class": "naruto-avatar naruto-avatar-circle",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "class": "naruto-icon naruto-icon-user",
            },
            "children": Array [
              Object {
                "attribs": Object {
                  "href": "#icon-user",
                },
                "children": Array [],
                "name": "use",
                "namespace": "http://www.w3.org/2000/svg",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "href": "http://www.w3.org/1999/xlink",
                },
                "x-attribsPrefix": Object {
                  "href": "xlink",
                },
              },
            ],
            "name": "svg",
            "namespace": "http://www.w3.org/2000/svg",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
        ],
        "name": "span",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Object {
          "attribs": Object {
            "class": "naruto-avatar naruto-avatar-circle",
            "style": "color:#f56a00;background-color:#fde3cf",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "class": "",
              },
              "children": Array [
                Object {
                  "data": "U",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "text",
                },
              ],
              "name": "span",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
          ],
          "name": "span",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": Object {
            "attribs": Object {
              "class": "naruto-avatar naruto-avatar-circle",
              "style": "background-color:#87d068",
            },
            "children": Array [
              Object {
                "attribs": Object {
                  "class": "naruto-icon naruto-icon-user",
                },
                "children": Array [
                  Object {
                    "attribs": Object {
                      "href": "#icon-user",
                    },
                    "children": Array [],
                    "name": "use",
                    "namespace": "http://www.w3.org/2000/svg",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "tag",
                    "x-attribsNamespace": Object {
                      "href": "http://www.w3.org/1999/xlink",
                    },
                    "x-attribsPrefix": Object {
                      "href": "xlink",
                    },
                  },
                ],
                "name": "svg",
                "namespace": "http://www.w3.org/2000/svg",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                },
              },
            ],
            "name": "span",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": [Circular],
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
              "style": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
              "style": undefined,
            },
          },
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
            "style": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
            "style": undefined,
          },
        },
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {
            "class": "naruto-avatar naruto-avatar-circle naruto-avatar-image",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "alt": "",
                "src": "https://zos.alipayobjects.com/rmsportal/ODTLcjxAfvqbxHnVXCYX.png",
                "srcset": "",
              },
              "children": Array [],
              "name": "img",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "alt": undefined,
                "src": undefined,
                "srcset": undefined,
              },
              "x-attribsPrefix": Object {
                "alt": undefined,
                "src": undefined,
                "srcset": undefined,
              },
            },
          ],
          "name": "span",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
      Object {
        "attribs": Object {
          "class": "naruto-avatar naruto-avatar-circle",
          "style": "color:#f56a00;background-color:#fde3cf",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "class": "",
            },
            "children": Array [
              Object {
                "data": "U",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "text",
              },
            ],
            "name": "span",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
        ],
        "name": "span",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Object {
          "attribs": Object {
            "class": "naruto-avatar naruto-avatar-circle",
            "style": "background-color:#87d068",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "class": "naruto-icon naruto-icon-user",
              },
              "children": Array [
                Object {
                  "attribs": Object {
                    "href": "#icon-user",
                  },
                  "children": Array [],
                  "name": "use",
                  "namespace": "http://www.w3.org/2000/svg",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "href": "http://www.w3.org/1999/xlink",
                  },
                  "x-attribsPrefix": Object {
                    "href": "xlink",
                  },
                },
              ],
              "name": "svg",
              "namespace": "http://www.w3.org/2000/svg",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
          ],
          "name": "span",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
            "style": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
            "style": undefined,
          },
        },
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {
            "class": "naruto-avatar naruto-avatar-circle",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "class": "naruto-icon naruto-icon-user",
              },
              "children": Array [
                Object {
                  "attribs": Object {
                    "href": "#icon-user",
                  },
                  "children": Array [],
                  "name": "use",
                  "namespace": "http://www.w3.org/2000/svg",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "href": "http://www.w3.org/1999/xlink",
                  },
                  "x-attribsPrefix": Object {
                    "href": "xlink",
                  },
                },
              ],
              "name": "svg",
              "namespace": "http://www.w3.org/2000/svg",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
          ],
          "name": "span",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": Object {
            "attribs": Object {
              "class": "naruto-avatar naruto-avatar-circle naruto-avatar-image",
            },
            "children": Array [
              Object {
                "attribs": Object {
                  "alt": "",
                  "src": "https://zos.alipayobjects.com/rmsportal/ODTLcjxAfvqbxHnVXCYX.png",
                  "srcset": "",
                },
                "children": Array [],
                "name": "img",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "alt": undefined,
                  "src": undefined,
                  "srcset": undefined,
                },
                "x-attribsPrefix": Object {
                  "alt": undefined,
                  "src": undefined,
                  "srcset": undefined,
                },
              },
            ],
            "name": "span",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": [Circular],
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
          "style": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
          "style": undefined,
        },
      },
      Object {
        "attribs": Object {
          "class": "naruto-avatar naruto-avatar-circle",
          "style": "background-color:#87d068",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "class": "naruto-icon naruto-icon-user",
            },
            "children": Array [
              Object {
                "attribs": Object {
                  "href": "#icon-user",
                },
                "children": Array [],
                "name": "use",
                "namespace": "http://www.w3.org/2000/svg",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "href": "http://www.w3.org/1999/xlink",
                },
                "x-attribsPrefix": Object {
                  "href": "xlink",
                },
              },
            ],
            "name": "svg",
            "namespace": "http://www.w3.org/2000/svg",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
        ],
        "name": "span",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {
            "class": "naruto-avatar naruto-avatar-circle",
            "style": "color:#f56a00;background-color:#fde3cf",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "class": "",
              },
              "children": Array [
                Object {
                  "data": "U",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "text",
                },
              ],
              "name": "span",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
          ],
          "name": "span",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": Object {
            "attribs": Object {
              "class": "naruto-avatar naruto-avatar-circle",
            },
            "children": Array [
              Object {
                "attribs": Object {
                  "class": "naruto-icon naruto-icon-user",
                },
                "children": Array [
                  Object {
                    "attribs": Object {
                      "href": "#icon-user",
                    },
                    "children": Array [],
                    "name": "use",
                    "namespace": "http://www.w3.org/2000/svg",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "tag",
                    "x-attribsNamespace": Object {
                      "href": "http://www.w3.org/1999/xlink",
                    },
                    "x-attribsPrefix": Object {
                      "href": "xlink",
                    },
                  },
                ],
                "name": "svg",
                "namespace": "http://www.w3.org/2000/svg",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                },
              },
            ],
            "name": "span",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": [Circular],
            "parent": [Circular],
            "prev": Object {
              "attribs": Object {
                "class": "naruto-avatar naruto-avatar-circle naruto-avatar-image",
              },
              "children": Array [
                Object {
                  "attribs": Object {
                    "alt": "",
                    "src": "https://zos.alipayobjects.com/rmsportal/ODTLcjxAfvqbxHnVXCYX.png",
                    "srcset": "",
                  },
                  "children": Array [],
                  "name": "img",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "alt": undefined,
                    "src": undefined,
                    "srcset": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "alt": undefined,
                    "src": undefined,
                    "srcset": undefined,
                  },
                },
              ],
              "name": "span",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": [Circular],
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
            "style": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
            "style": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
          "style": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
          "style": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {},
    "x-attribsPrefix": Object {},
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
